File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Petronite/PetroniteProtocolDtos.cs
using System.Text.Json.Serialization;

namespace FccDesktopAgent.Core.Adapter.Petronite;

/// <summary>
/// Token response from POST /oauth/token.
/// </summary>
public sealed record PetroniteTokenResponse(
    [property: JsonPropertyName("access_token")] string AccessToken,
    [property: JsonPropertyName("token_type")] string TokenType,
    [property: JsonPropertyName("expires_in")] int ExpiresIn);

/// <summary>
/// Nozzle assignment from GET /nozzles/assigned.
/// Maps a Petronite nozzle ID to canonical pump/nozzle numbers.
/// </summary>
public sealed record PetroniteNozzleAssignment(
    [property: JsonPropertyName("nozzleId")] string NozzleId,
    [property: JsonPropertyName("pumpNumber")] int PumpNumber,
    [property: JsonPropertyName("nozzleNumber")] int NozzleNumber,
    [property: JsonPropertyName("productCode")] string ProductCode,
    [property: JsonPropertyName("productName")] string? ProductName,
    [property: JsonPropertyName("status")] string Status);

/// <summary>
/// Create order request body for POST /direct-authorize-requests/create.
/// </summary>
public sealed record PetroniteCreateOrderRequest(
    [property: JsonPropertyName("nozzleId")] string NozzleId,
    [property: JsonPropertyName("maxVolumeLitres")] decimal MaxVolumeLitres,
    [property: JsonPropertyName("maxAmountMajor")] decimal MaxAmountMajor,
    [property: JsonPropertyName("currency")] string Currency,
    [property: JsonPropertyName("externalReference")] string? ExternalReference);

/// <summary>
/// Create order response from POST /direct-authorize-requests/create.
/// </summary>
public sealed record PetroniteCreateOrderResponse(
    [property: JsonPropertyName("orderId")] string OrderId,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("message")] string? Message);

/// <summary>
/// Authorize pump request for POST /direct-authorize-requests/authorize.
/// </summary>
public sealed record PetroniteAuthorizeRequest(
    [property: JsonPropertyName("orderId")] string OrderId);

/// <summary>
/// Authorize pump response from POST /direct-authorize-requests/authorize.
/// </summary>
public sealed record PetroniteAuthorizeResponse(
    [property: JsonPropertyName("orderId")] string OrderId,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("authorizationCode")] string? AuthorizationCode,
    [property: JsonPropertyName("message")] string? Message);

/// <summary>
/// Webhook payload (POST from Petronite to our webhook endpoint).
/// </summary>
public sealed record PetroniteWebhookPayload(
    [property: JsonPropertyName("eventType")] string EventType,
    [property: JsonPropertyName("transaction")] PetroniteTransactionData? Transaction,
    [property: JsonPropertyName("timestamp")] string Timestamp);

/// <summary>
/// Transaction data within a Petronite webhook payload.
/// Volume and amounts are in major units (decimal) — the adapter converts to canonical minor/micro units.
/// </summary>
public sealed record PetroniteTransactionData(
    [property: JsonPropertyName("orderId")] string OrderId,
    [property: JsonPropertyName("nozzleId")] string NozzleId,
    [property: JsonPropertyName("pumpNumber")] int PumpNumber,
    [property: JsonPropertyName("nozzleNumber")] int NozzleNumber,
    [property: JsonPropertyName("productCode")] string ProductCode,
    [property: JsonPropertyName("volumeLitres")] decimal VolumeLitres,
    [property: JsonPropertyName("amountMajor")] decimal AmountMajor,
    [property: JsonPropertyName("unitPrice")] decimal UnitPrice,
    [property: JsonPropertyName("currency")] string Currency,
    [property: JsonPropertyName("startTime")] string StartTime,
    [property: JsonPropertyName("endTime")] string EndTime,
    [property: JsonPropertyName("receiptCode")] string? ReceiptCode,
    [property: JsonPropertyName("attendantId")] string? AttendantId,
    [property: JsonPropertyName("paymentMethod")] string PaymentMethod);

/// <summary>
/// Pending order from GET /direct-authorize-requests/pending.
/// </summary>
public sealed record PetronitePendingOrder(
    [property: JsonPropertyName("orderId")] string OrderId,
    [property: JsonPropertyName("nozzleId")] string NozzleId,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("createdAt")] string CreatedAt,
    [property: JsonPropertyName("maxVolumeLitres")] decimal? MaxVolumeLitres,
    [property: JsonPropertyName("maxAmountMajor")] decimal? MaxAmountMajor);

/// <summary>
/// Field-level validation error returned by the Petronite API.
/// </summary>
public sealed record PetroniteFieldError(
    [property: JsonPropertyName("field")] string Field,
    [property: JsonPropertyName("message")] string Message);

/// <summary>
/// Error response wrapper returned by the Petronite API on non-2xx responses.
/// </summary>
public sealed record PetroniteErrorResponse(
    [property: JsonPropertyName("errorCode")] string ErrorCode,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("errors")] IReadOnlyList<PetroniteFieldError>? Errors);

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/petronite/PetroniteProtocolDtos.kt
package com.fccmiddleware.edge.adapter.petronite

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ---------------------------------------------------------------------------
// Petronite protocol DTOs — JSON structures for Petronite REST API
// ---------------------------------------------------------------------------

/**
 * Token response from POST /oauth/token.
 */
@Serializable
data class PetroniteTokenResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("token_type") val tokenType: String,
    @SerialName("expires_in") val expiresIn: Int,
)

/**
 * Nozzle assignment from GET /nozzles/assigned.
 * Maps a Petronite nozzle ID to canonical pump/nozzle numbers.
 */
@Serializable
data class PetroniteNozzleAssignment(
    @SerialName("nozzleId") val nozzleId: String,
    @SerialName("pumpNumber") val pumpNumber: Int,
    @SerialName("nozzleNumber") val nozzleNumber: Int,
    @SerialName("productCode") val productCode: String,
    @SerialName("productName") val productName: String? = null,
    @SerialName("status") val status: String,
)

/**
 * Create order request body for POST /direct-authorize-requests/create.
 */
@Serializable
data class PetroniteCreateOrderRequest(
    @SerialName("nozzleId") val nozzleId: String,
    @SerialName("maxVolumeLitres") val maxVolumeLitres: String,
    @SerialName("maxAmountMajor") val maxAmountMajor: String,
    @SerialName("currency") val currency: String,
    @SerialName("externalReference") val externalReference: String? = null,
)

/**
 * Create order response from POST /direct-authorize-requests/create.
 */
@Serializable
data class PetroniteCreateOrderResponse(
    @SerialName("orderId") val orderId: String,
    @SerialName("status") val status: String,
    @SerialName("message") val message: String? = null,
)

/**
 * Authorize pump request for POST /direct-authorize-requests/authorize.
 */
@Serializable
data class PetroniteAuthorizeRequest(
    @SerialName("orderId") val orderId: String,
)

/**
 * Authorize pump response from POST /direct-authorize-requests/authorize.
 */
@Serializable
data class PetroniteAuthorizeResponse(
    @SerialName("orderId") val orderId: String,
    @SerialName("status") val status: String,
    @SerialName("authorizationCode") val authorizationCode: String? = null,
    @SerialName("message") val message: String? = null,
)

/**
 * Webhook payload (POST from Petronite to our webhook endpoint).
 */
@Serializable
data class PetroniteWebhookPayload(
    @SerialName("eventType") val eventType: String,
    @SerialName("transaction") val transaction: PetroniteTransactionData? = null,
    @SerialName("timestamp") val timestamp: String,
)

/**
 * Transaction data within a Petronite webhook payload.
 * Volume and amounts are in major units (decimal strings) -- the adapter converts
 * to canonical minor/micro units using BigDecimal.
 */
@Serializable
data class PetroniteTransactionData(
    @SerialName("orderId") val orderId: String,
    @SerialName("nozzleId") val nozzleId: String,
    @SerialName("pumpNumber") val pumpNumber: Int,
    @SerialName("nozzleNumber") val nozzleNumber: Int,
    @SerialName("productCode") val productCode: String,
    @SerialName("volumeLitres") val volumeLitres: String,
    @SerialName("amountMajor") val amountMajor: String,
    @SerialName("unitPrice") val unitPrice: String,
    @SerialName("currency") val currency: String,
    @SerialName("startTime") val startTime: String,
    @SerialName("endTime") val endTime: String,
    @SerialName("receiptCode") val receiptCode: String? = null,
    @SerialName("attendantId") val attendantId: String? = null,
    @SerialName("paymentMethod") val paymentMethod: String,
)

/**
 * Pending order from GET /direct-authorize-requests/pending.
 */
@Serializable
data class PetronitePendingOrder(
    @SerialName("orderId") val orderId: String,
    @SerialName("nozzleId") val nozzleId: String,
    @SerialName("status") val status: String,
    @SerialName("createdAt") val createdAt: String,
    @SerialName("maxVolumeLitres") val maxVolumeLitres: String? = null,
    @SerialName("maxAmountMajor") val maxAmountMajor: String? = null,
)

/**
 * Field-level validation error returned by the Petronite API.
 */
@Serializable
data class PetroniteFieldError(
    @SerialName("field") val field: String,
    @SerialName("message") val message: String,
)

/**
 * Error response wrapper returned by the Petronite API on non-2xx responses.
 */
@Serializable
data class PetroniteErrorResponse(
    @SerialName("errorCode") val errorCode: String,
    @SerialName("message") val message: String,
    @SerialName("errors") val errors: List<PetroniteFieldError>? = null,
)

/**
 * Internal record tracking an active pre-authorization for correlation with
 * incoming webhook transactions.
 */
data class ActivePreAuth(
    /** Petronite-assigned order ID. */
    val orderId: String,
    /** Petronite nozzle ID used for the pre-auth. */
    val nozzleId: String,
    /** Canonical pump number from the original PreAuthCommand. */
    val pumpNumber: Int,
    /** Odoo order ID for correlation, if provided. */
    val odooOrderId: String?,
    /** Pre-auth record ID from the original command, if provided. */
    val preAuthId: String?,
    /** When this pre-auth was created. UTC epoch millis. */
    val createdAtMillis: Long,
)
